### PR TITLE
[Merged by Bors] - bevy_ecs: Use 32-bit entity ID cursor on platforms without AtomicI64

### LINKED ISSUE #4451: bevy_ecs: Allow using i32 for entity ID cursor on platforms without AtomicI64 support
## What problem does this solve or what need does it fill?

I have been trying to get bits and pieces of Bevy building for an unusual target platform (`armv6k-nintendo-3ds`), and to my surprise `bevy_ecs` almost compiles already!

The only issue I ran into was in https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/entity/mod.rs#L35 – since the platform does not have atomic load/stores for `i64`, there is no `AtomicI64` available for import. I imagine that some other 32-bit platforms may also run into this problem, but have probably not been the focus of the project thus far.

## What solution would you like?

With Rust 1.60.0 the `#[cfg(target_has_atomic = "64")]` flag has just been stabilized: https://github.com/rust-lang/rust/pull/93824. I propose using this to conditionally compile using `AtomicI32` on platforms that do not support 64-bit atomics.

I'm not sure if there might be any further-reaching consequences of such a change, but this platform is already not supported, so it seems like adding support in this way would still be "experimental" or possibly unstable, but still an improvement on the status quo. Existing platforms' implementation would not be impacted.

## What alternative(s) have you considered?

No action; effectively require using a fork for platforms without atomic i64 support.

## Additional context

I already have a local change with which I have successfully built for `armv6k-nintendo-3ds` – I'm happy to open a PR if this proposed solution is accepted.

### PR DESCRIPTION
**Pull request by ian-h-chamberlain**
# Objective
- Fixes #4451

## Solution
- Conditionally compile entity ID cursor as `AtomicI32` when compiling on a platform that does not support 64-bit atomics.

- This effectively raises the MSRV to 1.60 as it uses a `#[cfg]` that was only just stabilized there. (should this be noted in changelog?)

---

## Changelog
- Added `bevy_ecs` support for platforms without 64-bit atomic ints

## Migration Guide
N/A

### CODE REVIEW

**Review by TheRawMeatball — APPROVED** [review]

File: crates/bevy_ecs/src/entity/mod.rs
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:37]: Not a huge fan of the name `Int` for this, but using a typedef here makes sense.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:259]: This has wrapping behaviour on overflow, which means that we don't fail gracefully when the `free_cursor` overflows.
On the other hand, that requires `2^31` items to be requested between flushes, which does seem unlikely.
Luckily, we just panic further below at the indexing in that case, so it's not the end of the world.
If you get to ridiculous numbers, like requesting `u32::MAX-k`, I think the subtraction below would also wrap in release, potentially leading to some very strange results.

Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:259]: This has saturating behaviour on overflow, which means that we don't fail gracefully when `count` overflows an `i32`.
This one is potentially worse, since it's possible we rely on `ReserveEntitiesIterator` to give exactly the requested number of entities somewhere.
```suggestion
        let range_end = self.free_cursor.fetch_sub(count.try_into().unwrap(), Ordering::Relaxed);
```
which is semantically identical as the current behaviour on normal platforms, but avoids this pitfall. This might also fix the other comment, but I haven't thought this through properly

File: crates/bevy_ecs/src/entity/mod.rs
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:40]: In https://github.com/bevyengine/bevy/pull/2827, we decided to use `AtomicUsize` for `WorldIds`, since that is guaranteed to be available if you can get 'normal' `std` compiling.
Would it make sense to use `AtomicIsize` here on platforms without `AtomicI64` support - that is, once we work out all the wrapping behaviours properly?

File: crates/bevy_ecs/src/entity/mod.rs
Comment (DJMcNab) [inline crates/bevy_ecs/src/entity/mod.rs:35]: This should have a comment for posterity, mentioning that all mainstream platforms do make 64 bit atomics available, but this technique adds support for them with some limitations.

**Review by DJMcNab — COMMENTED** [review]
It's also worth noting that (as far as I know) we don't plan to make any guarantees about weird platforms at this point in time. That is, if you want it to keep working, you'll likely need to keep patching yourself. 

Additionally, I wouldn't be super surprised if this didn't land. I'd say that this probably shouldn't be merged without cart's sign off, since it is definitely a weird hack.

That being said, I'm reasonably  happy with the implementation of the change, barring the few corner cases which need additional thought applied.

**Comment by ian-h-chamberlain** [comment]
> It's also worth noting that (as far as I know) we don't plan to make any guarantees about weird platforms at this point in time. That is, if you want it to keep working, you'll likely need to keep patching yourself.
> 
> Additionally, I wouldn't be super surprised if this didn't land. I'd say that this probably shouldn't be merged without cart's sign off, since it is definitely a weird hack.

Yeah, I figured this might be the case, but the change was relatively small in scope so I thought I'd see if it would be accepted into mainline. I realize it's not a stated goal of the project to maintain compatibility for unusual platforms like this, but if there is a commit hash I can pin to and use Bevy with I will be pretty happy with that :)

I will look to see if I can at least address your initial concerns / comments and we can go from there (if/when cart has a chance to look over it as well).

File: crates/bevy_ecs/src/entity/mod.rs
Comment (ian-h-chamberlain) [reply crates/bevy_ecs/src/entity/mod.rs:37]: Yeah, I didn't love it either, some alternatives might be
* `IdCursor` or something like this?
* `NonAtomicInt` - feels a little heavy-handed
* just `int`, although this might get flagged by clippy or something, but is more in line with e.g. `libc` conventions for types like this

Open to other suggestions as well.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (ian-h-chamberlain) [reply crates/bevy_ecs/src/entity/mod.rs:259]: Ok, this is the kind of thing I was unsure of the impact of a change like this. Would the "proper" solution be to drop ID sizes down to `u16` or something to avoid this, or is this wrapping behavior simply a caveat that would need to be documented for these kinds of platforms?

File: crates/bevy_ecs/src/entity/mod.rs
Comment (ian-h-chamberlain) [reply crates/bevy_ecs/src/entity/mod.rs:40]: I think `AtomicIsize` is fine to use as the fallback here, although like you mention the difference probably only matters for 16-bit platforms. If it might increase compatibility for such platforms then I suppose it might as well use `isize` rather than `i32` – although as you point out it may be moot if there are other blockers on such platforms.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/entity/mod.rs:259]: I wonder if we can / should fix this in general 🤔 If so, we should likely split it out into its own PR though.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/entity/mod.rs:259]: Per https://github.com/bevyengine/bevy/issues/3899#issuecomment-1094361070, I think we should drop comments above this unwrap explaining why this could panic.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/entity/mod.rs:37]: I prefer `IdCursor` or something to that effect. It's very much used for a single purpose here, and "int" and the like imply a lot more generality.

**Comment by alice-i-cecile** [comment]
> This effectively raises the MSRV to 1.60 as it uses a #[cfg] that was only just stabilized there. (should this be noted in changelog?)

Our stated latest supported release is "latest stable", so no worries there :)

This is an interesting and not very intrusive change 🤔 Needs some cleanup, but I personally don't mind including small tweaks  like this for compatibility if they're well-documented and unobtrusive. Well down the road, I think we should consider something like Rust's platform tiers.

You could obviously maintain a fork of Bevy with these minimal changes, but that becomes rather frustrating in the context of the ecosystem, so it would be nice to be able to upstream this.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (DJMcNab) [reply crates/bevy_ecs/src/entity/mod.rs:259]: For the other platforms, since `i64` is so much bigger than `u32`, if too many entities are requested, we should always panic at the `try_from` below, as far as I can tell.

**Comment by ian-h-chamberlain** [comment]
Ok, thanks for the helpful feedback! Just to clarify, am I missing anything requested for changes here?

- [ ] Rename type alias to something more descriptive (probably `IdCursor`)
- [ ] Document the fallback type alias as a compatibility shim
- [ ] Use `AtomicIsize` instead of `AtomicI32` in the non-64-bit case?
- [ ] Use `try_from().unwrap()` in cases where casting may overflow, and add unwrap comment explaining how it can fail

Is that everything? It sounded like maybe the unwrap could go into a different PR, but maybe it it belongs in this one since the risk of overflow is greater on these kinds of platforms?

**Comment by alice-i-cecile** [comment]
Yep, that sounds correct :) Final call on whether we want this sort of compatibility change at all belongs to @cart as a matter of controversial engine-level policy, but I think that those changes should make this PR  mergeable if he decides to pry open that particular can of worms.

Just to double-check @ian-h-chamberlain: this was the only change you needed to get `bevy_ecs` compiling on a 32 bit platform?

**Comment by ian-h-chamberlain** [comment]
> Just to double-check @ian-h-chamberlain: this was the only change you needed to get `bevy_ecs` compiling on a 32 bit platform?

@alice-i-cecile yes, actually with this change and a custom nightly I was able to compile the whole `bevy` crate with `default-features = false`! My plan was to try enabling relevant features and see what else might be worth upstreaming if I found issues, but this seemed like a good starting point.

I have only tested on the `armv6k-nintendo-3ds` target, which has a (WIP + unstable) `std` implementation, so I'm not sure how well it would work for other 32-bit Rust-tier-3 targets, but if they have `std` then probably it would work?

----

I'll try to update the PR later this evening with the requested changes.

**Comment by ian-h-chamberlain** [comment]
Updated with the requested changes, let me know if there's anything else needed!

File: crates/bevy_ecs/src/entity/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/entity/mod.rs:264]: This comment looks like it's a line too high?

**Review by alice-i-cecile — APPROVED** [review]
Code LGTM now. Still going to punt on the "should we do this" call though :)

**Comment by ian-h-chamberlain** [comment]
@alice-i-cecile Based on the recent Discord announcement regarding "controversial" PRs (and your previous comments), should this be labeled with `S-Controversial`? I would presume that you'd still want cart's approval on this, even if other maintainers signed off on this change?

**Comment by alice-i-cecile** [comment]
Good call, thanks.

**Review by IceSentry — APPROVED** [review]
The side effects are a bit controversial, but the code looks good to me.

**Comment by ian-h-chamberlain** [comment]
@alice-i-cecile I think this has passed the 45-day mark since you marked it as ready for review, is there anything else (a final pass of reviews maybe) needed to get it merged?

**Comment by alice-i-cecile** [comment]
For the 45 day mark, I needed a second approval from either @mockersf or @robswain. I think this is a good candidate for that process though; this is controversial but low stakes and will basically never be the "most important thing to review" :)

**Comment by robswain** [comment]
@alice-i-cecile You've got the wrong Rob Swain I'm afraid. I definitely don't have anything to do with this project 😄

**Comment by bjorn3** [comment]
Indeed. cc @superdump

**Comment by mockersf** [comment]
This PR looks good, but there is no way to guarantee it's still working. Another PR could break the support and we would never know until someone complains.
Is there a reason to not use `isize`/`AtomicIsize` everywhere?

What kind of guarantee do we want to provide? @ian-h-chamberlain would you be happy with having to fix it occasionally? Do we want to setup a CI job that would cross compile to a target without 64 bits atomics?

**Comment by ian-h-chamberlain** [comment]
In terms of support – for the `armv6k-nintendo-3ds` target (or I suppose any other tier 3 targets that happen to benefit from this change), I wouldn't expect Bevy to maintain compatibility continuously, and I'd be fine with fixing it whenever I experienced breakage.

However, it might be reasonable to add CI for a tier 1 target like `i686-unknown-linux-gnu`. @mockersf would you envision something like `.github/workflows/ios.yml` that runs periodically, or something as part of the main CI pipeline (like `build-wasm` and `build-android`)? I can look into this, I think ubuntu runners should be able to cross compile to a target like this without too much difficulty.

**Comment by DJMcNab** [comment]
> Is there a reason to not use isize/AtomicIsize everywhere?

We want the smallest signed integer big enough to contain `u32::MAX`.

That is `i64`. It just so happens that some platforms don't support 64 bit atomics, so for those rare cases, it's better to fall back on isize, which (since we use at least 4 bytes of overhead per entity) is definitely big enough to contain as many entities as can be spawned (`u32::MAX`, but capped at `usize::MAX/4`).

The reason not to use isize everywhere is that it's wasteful on 128 bit platforms. Admittedly, I don't know of any 128 bit platforms, but I'd rather keep it as the condition explained in my first paragraph.

And I think the consensus is that we don't guarantee it works. We just don't intentionally stand in the way of it not working (without good reason, anyway). In rustc parlance, these are tier 3 targets, i.e. it might be supported, and you can bring PRs to fix the support, but we don't try and ensure it is.

**Comment by mockersf** [comment]
Thanks for the explanation!

> @mockersf would you envision something like .github/workflows/ios.yml that runs periodically, or something as part of the main CI pipeline (like build-wasm and build-android)?

If GitHub had better support for actions on cron, maybe, but as it is for the moment it's pretty useless. And I would rather not add it to the main CI jobs, they are already long enough.
The best from my point of view is if you had a project somewhere that is tested regularly on your target architecture and would report issues, but that put the burden on you.

**Review by mockersf — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Yep, we're on board with that setup for now. This is still controversial though, so I've started the timer to merge :)

**Comment by superdump** [comment]
> @alice-i-cecile You've got the wrong Rob Swain I'm afraid. I definitely don't have anything to do with this project 😄

Hello name-sharer! Much love to you!

**Review by superdump — APPROVED** [review]

**Comment by ian-h-chamberlain** [comment]
It looks like this has reached the 45-day mark, based on the project page. Is there anything else needed to merge now that we've hit the timer?

**Comment by alice-i-cecile** [comment]
I have no further objections :) I'm on vacation though, so I'll merge this at the start of September if no one else beats me to it.

**Comment by cart** [comment]
If this change was any more invasive than it is, I'd probably block it. But its _just_ scoped enough that I'm willing to squeeze it in. It makes me just a little bit uncomfortable and we're catering to a pretty niche crowd here (3ds is effectively dead at this point, so this is enabling homebrew stuff? And I haven't seen this issue come up on any other platforms). Getting bevy working on a 3ds is definitely fun and cool though, and we should encourage doing fun and cool things, within reason.

But for the person (or AI) years in the future reading this thread and writing future Bevy ECS code: if you need to drop this support for any legitimate reason ... I think you should do it.

**Pull request closed without merging**
